src/mongocxx/test/spec/entity.hh

```diff
@@ -0,0 +1,57 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongocxx/config/private/prelude.hh>
+
+#include <unordered_map>
+
+#include <bsoncxx/stdx/variant.hpp>
+#include <mongocxx/client.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+namespace entity {
+
+class map {
+   public:
+    using key_type = mongocxx::stdx::string_view;
+    using mapped_type = bsoncxx::stdx::variant<mongocxx::client,
+                                               mongocxx::database,
+                                               mongocxx::collection,
+                                               mongocxx::client_session,
+                                               mongocxx::gridfs::bucket>;
+
+    map() = default;
+
+    map(const map&) = delete;
+    map operator=(const map&) = delete;
+
+    map(map&&) = default;
+    map& operator=(map&&) = default;
+
+    void insert(const key_type& key, mongocxx::client&& client);
```

Comment:
Will add insert methods for the other entities as well. The idea here was to encapsulate the variant and have the entity class worry about type safety. There are no getters yet. We won't need them until we run operations. The getters will return mongocxx types (client, db, coll, etc) given a key rather than mapped_type.

Reply:
Nice, I assume `insert` and `get` could be templated?

Reply:
Yep. This class shouldn't get much larger than it is now after adding support for the other entities.